# Regulated IRE1α-dependent decay (RIDD) mediated reprograming of lipid metabolism in cancer

This repository contains materials and methods for the paper "Regulated IRE1α-dependent decay (RIDD)-mediated reprograming of lipid metabolism in cancer" by Almanza*, Mnich* et al., (2021)

<!-- ABOUT THE PROJECT -->
## About The Project

In this study, we employed unbiased lipidomic and transcriptomic approaches to investigate the contribution of IRE1α signalling to lipid metabolism in TNBC. This system biology framework enables the systematic exploration of regulation of metabolic pathways by IRE1α and represents a novel resource for the study of non-canonical IRE1α functions.

<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary>Table of Contents</summary>
  <ol>
    <li><a href="#Data">Data</a></li>
    <li><a href="#Scripts">Scripts</a></li>
    <li><a href="#Figures">Figures</a></li>
    <li><a href="#Tables">Tables</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>

<!-- DATA -->
## Data

The following datasets are included in the repository:

* Metasyx_MD.txt contains metabolomics measurement day normalized intensities of detected features (annotated and non-annotated). 30 samples that were exposed to the  IRE1α inhibitor or DMSO. Samples were collected at 24, 48 and 72h after the treatment. Five replicates were measured per each condition.

* Metasyx_Sample_Info.txt contains information about group structure and measurement order.

* raw_counts.txt Raw RNAseq counts from featureCounts. 12 samples that were exposed to the IRE1α inhibitor or DMSO. Samples were collected at 8 and 24h after the treatment. Three replicates measured per each condition.

* sample_info.txt contains sample information for the RNAseq experiment.

* 

<!-- SCRIPTS -->
## Scripts

* edgeR_final.v.R: : performs the differential gene expression (DEG) analysis of the transcriptomic data in the paper. It generates S3 b,c,d,e figures. 

*


<!-- FIGURES -->
## Figures


<!-- TABLES -->
## Tables



<!-- CONTACT -->
## Contact




